quesma/queryparser/pancake_pipelines.go

```diff
@@ -0,0 +1,178 @@
+	subAggrRows [][]model.QueryResultRow) (resultRowsPerPipeline map[string][]model.QueryResultRow) {
+
+	resultRowsPerPipeline = make(map[string][]model.QueryResultRow)
+
+	for _, childPipeline := range nextLayer.childrenPipelineAggregations {
+		if childPipeline.queryType.PipelineAggregationType() != model.PipelineBucketAggregation {
+			continue
+		}
+
+		needToAddProperMetricColumn := !childPipeline.queryType.IsCount() // If count, last column of bucketRows is already count we need.
+		if needToAddProperMetricColumn {
+			bucketRows = p.addProperPipelineColumn(childPipeline.parentInternalName, bucketRows, subAggrRows)
+		}
+
+		var bucketRowsTransformedIfNeeded []model.QueryResultRow
+		switch queryType := layer.nextBucketAggregation.queryType.(type) {
+		// TODO: logic what and when to transform shouldn't probably be here
+		case bucket_aggregations.Histogram:
+			bucketRowsTransformedIfNeeded = queryType.NewRowsTransformer().Transform(p.ctx, bucketRows)
+		case *bucket_aggregations.DateHistogram:
+			bucketRowsTransformedIfNeeded = queryType.NewRowsTransformer().Transform(p.ctx, bucketRows)
+		default:
+			bucketRowsTransformedIfNeeded = bucketRows
+		}
+
+		childResults := p.calcSinglePipelineBucket(nextLayer, childPipeline, bucketRowsTransformedIfNeeded)
+		for pipelineName, pipelineResults := range childResults {
+			if _, alreadyExists := resultRowsPerPipeline[pipelineName]; alreadyExists { // sanity check
+				logger.ErrorWithCtx(p.ctx).Msgf("pipeline %s already exists in resultsPerPipeline", pipelineName)
+			}
+			resultRowsPerPipeline[pipelineName] = pipelineResults
+		}
+	}
+
+	return
+}
+
+func (p pancakePipelinesProcessor) calcSinglePipelineBucket(layer *pancakeModelLayer, pipeline *pancakeModelPipelineAggregation,
+	bucketRows []model.QueryResultRow) (resultRowsPerPipeline map[string][]model.QueryResultRow) {
+
+	resultRowsPerPipeline = make(map[string][]model.QueryResultRow)
+
+	currentPipelineResults := pipeline.queryType.CalculateResultWhenMissing(bucketRows)
+	resultRowsPerPipeline[pipeline.name] = currentPipelineResults
+
+	for _, pipelineChild := range layer.findPipelineChildren(pipeline) {
+		childPipelineResults := p.calcSinglePipelineBucket(layer, pipelineChild, currentPipelineResults)
+		for name, results := range childPipelineResults {
+			if _, alreadyExists := resultRowsPerPipeline[name]; alreadyExists { // sanity check
+				logger.ErrorWithCtx(p.ctx).Msgf("pipeline %s already exists in resultsPerPipeline", name)
+			}
+			resultRowsPerPipeline[name] = results
+		}
+	}
+
+	return
+}
+
+// returns:
+//   - newSelectedRows: same as selectedRows, but with one column different if needed (value for this column is taken from
+//     allRows, which has >= columns than selectedRows, and should have the column we need)
+//   - oldColumnArray:  old value of the exchanged column, to be restored in restoreOriginalColumn after processing
+//
+// Use restoreOriginalColumn after processing to restore original values.
```

Comment:
The comment needs update.

Still I think we may be able to avoid this method at all, though need to experiment myself.